algebra-core/src/curves/cuda/scalar_mul/cpu_gpu_macros.rs

```diff
@@ -0,0 +1,246 @@
+                let n_cpu = n - n_gpus.iter().sum::<usize>();
+
+                // Create storage for buffers and contexts for variable number of devices
+                let mut bases_split = Vec::with_capacity(n_devices);
+                let mut tables = Vec::with_capacity(n_devices);
+                let mut exps = Vec::with_capacity(n_devices);
+                let mut ctxs = Vec::with_capacity(n_devices);
+                let (mut time_cpu, mut times_gpu) = (0, vec![0; n_devices]);
+
+                // Split data and generate tables and u8 scalar encoding in device memory
+                for (i, &num) in n_gpus.iter().enumerate() {
+                    let device = Device::nth(i).unwrap();
+                    let ctx = device.create_context();
+
+                    let (lower, upper) = res_ref.split_at_mut(num);
+                    res_ref = upper;
+                    let lower_exps = &exps_h_ref[..num];
+                    exps_h_ref = &exps_h_ref[num..];
+
+                    let mut table = DeviceMemory::<Self>::zeros(&ctx, num * Self::table_size());
+                    let mut exp = DeviceMemory::<u8>::zeros(&ctx, num * Self::num_u8());
+
+                    Self::generate_tables_and_recoding(lower, &mut table[..], lower_exps, &mut exp[..]);
+
+                    ctxs.push((device, ctx));
+                    bases_split.push(lower);
+                    tables.push(table);
+                    exps.push(exp);
+                };
+
+                println!("Split statically and allocated device: {}us", now.elapsed().as_micros());
+
+                rayon::scope(|s| {
+                    // Run jobs on GPUs
+                    for (i, (bases_gpu, time_gpu)) in bases_split.iter_mut().zip(times_gpu.iter_mut()).enumerate() {
+                        let n_gpu = n_gpus[i];
+                        let ctx = &ctxs[i].1;
+                        let table = &tables[i];
+                        let exp = &exps[i];
+
+                        s.spawn(move |_| {
+                            let now = std::time::Instant::now();
+                            let mut out = DeviceMemory::<Self>::zeros(ctx, n_gpu);
+                            P::scalar_mul_kernel(
+                                ctx,
+                                (n_gpu - 1) / cuda_group_size + 1, // grid
+                                cuda_group_size,     // block
+                                table.as_ptr(), exp.as_ptr(), out.as_mut_ptr(), n_gpu as isize
+                            )
+                            .expect("Kernel call failed");
+                            Self::batch_normalization(&mut out[..]);
+                            bases_gpu.clone_from_slice(&out.par_iter().map(|p| p.into_affine()).collect::<Vec<_>>()[..]);
+                            *time_gpu = now.elapsed().as_micros();
+                            println!("GPU {} finish", i);
```

Comment:
This shouldn't appear in a normal run. Either change this to a debug print with logging or remove it.